docs/equality.md

```diff
@@ -0,0 +1,348 @@
+
+Here we define the relevant static type `EQTYPE` for the different constructs above:
+
+#### Basics
+
+* `a = b`:  `EQTYPE` is the statically known type of `a` or `b`
+* `a <> b`: `EQTYPE` is the statically known type of `a` or `b`
+
+#### Inlined constructs
+
+* `HashIdentity.Structural<'T>`,  EQTYPE is the **inlined** `'T` (results in specialized equality)
+* `Array.contains<'T>`,  EQTYPE is the **inlined** `'T` (results in specialized equality)
+* `List.contains<T>` likewise
+* `Seq.contains<T>` likewise
+
+These only resulting in naked generic equality if themselves used from a non-inlined generic context.
+
+#### Non-inlined constructs always resulting in naked generic equality
+
+* `Array.groupBy<'Key, 'T> f array`, EQTYPE is non-inlined 'Key, results in naked generic equality
+* `Array.countBy array` likewise for `'T`
+* `Array.distinct<'T> array` likewise
+* `Array.distinctBy array` likewise
+* `Array.except array` likewise
+* `List.groupBy` likewise
+* `List.countBy` likewise
+* `List.distinct` likewise
+* `List.distinctBy` likewise
+* `List.except` likewise
+* `Seq.groupBy` likewise
+* `Seq.countBy` likewise
+* `Seq.distinct` likewise
+* `Seq.distinctBy` likewise
+* `Seq.except` likewise
+* ... may be more here, see `: equality` constraints in `FSharp.Core` for non-inlined code. 
+
+These **always** result in naked generic equality checks.
+
+Example 1:
+
+```fsharp
+let x = HashIdentity.Structural<byte>  // EQTYPE known to compiler is `byte`
+```
+
+Example 2 (a non-inlined "naked" generic context):
+
+```fsharp
+let f2<'T> () =
+   ... some long code
+   // EQTYPE known to the compiler is `'T`
+   // RUNTIME-EQTYPE known to the library is `byte`
+   let x = HashIdentity.Structural<'T>
+   ... some long code
+
+f2<byte>() // performance of this determined by EQTYPE<'T> and RUNTIME-EQTYPE<byte>
+```
+
+Example 3 (an inlined generic context):
+
+```fsharp
+let f3<'T> () =
+   ... some long code
+   // EQTYPE known to the compiler is `byte`
+   // RUNTIME-EQTYPE known to the library is `byte`
+   let x = HashIdentity.Structural<'T>
+   ... some long code
+
+f3<byte>() // performance of this determined by EQTYPE<'T> and RUNTIME-EQTYPE<byte>
+```
+
+Example 4 (a generic struct type in a non-inline generic context):
+
+```fsharp
+let f4<'T> () =
+   ... some long code
+   // EQTYPE known to the compiler is `SomeStructType<'T>`
+   // RUNTIME-EQTYPE known to the library is `SomeStructType<byte>`
+   let x = HashIdentity.Structural<SomeStructType<'T>>
+   ... some long code
+
+f4<byte>() // performance of this determined by EQTYPE<SomeStructType<'T>> and RUNTIME-EQTYPE<SomeStructType<byte>>
+```
+
+## How we compile equality "a = b"
+
+This very much depends on the `EQTYPE` involved in the equality as known by the compiler
+
+Aim here is to flesh these all out with:
+* **Semantics**: what semantics the user expects, and what the semantics actually is
+* **Perf expectation**: what perf the user expects
```

Comment:
This seems superfluous, as all of explanations could be reduced to a single overarching statement about expecting the same perf as if hand-rolling manual equality.

Reply:
Well it goes a bit different in some cases, so will keep it.